src/main/scala/org/virtuslab/internal/load/parse/ParserImpl.scala

```diff
@@ -39,68 +39,96 @@ private enum Production:
 object ParserImpl extends Parser:
   import Production.*
 
-  def getEvents(in: Reader): Either[YamlError, List[Event]] =
+  override def getEvents(in: Reader): Either[YamlError, List[Event]] =
     @tailrec
-    def loop(acc: List[Event], productionStack: List[Production]): List[Event] = {
-      val (event, next) = ParserImpl.getNextEvent(in, productionStack)
-      if event != Event.StreamEnd then loop(acc :+ event, next)
-      else acc :+ event
+    def loop(
+        acc: List[Event],
+        productionStack: List[Production]
+    ): Either[YamlError, List[Event]] = {
+      ParserImpl.getNextEvent(in, productionStack) match
+        case Right((event, next)) =>
+          if event != Event.StreamEnd then loop(acc :+ event, next)
+          else Right(acc :+ event)
+        case Left(err) => Left(err)
     }
-    Right(loop(Nil, List(ParseStreamStart)))
+    loop(Nil, List(ParseStreamStart))
 
-  def getNextEvent(in: Reader, stack: List[Production]): (Event, List[Production]) = {
+  type EventResult = Either[YamlError, (Event, List[Production])]
+
+  private def error(expected: String, got: Token): EventResult = Left(
```

Comment:
We should separate the errors logic from the parsing. We should cover error parsing with new type error. 
In new type error, you can create helper method for parsing, such as 

`EventError.from(expected: Token, got: Token): EventResult`

Reply:
Should we change `YamlError` to sealed trait and then, `EventError` will extend it?

Reply:
Yes, good idea